src/bluesky/super_state_machine/machines.py

```diff
@@ -0,0 +1,275 @@
+"""State machine core."""
+
+from enum import Enum
+from functools import partial
+
+from . import utils
+
+NotSet = object()
+
+
+class DefaultMeta:
+    """Default configuration values."""
+
+    states_enum_name = "States"
+
+
+class AttributeDict(dict):
+    def __setattr__(self, key, value):
+        self[key] = value
+
+    def __getattr__(self, key):
+        return self[key]
+
+
+class StateMachineMetaclass(type):
+    """Metaclass for state machine, to build all its logic."""
+
+    def __new__(cls, name, bases, attrs):
+        """Create state machine and add all logic and methods to it."""
+        cls._set_up_context()
+        new_class = super(cls, cls).__new__(cls, name, bases, attrs)
+        cls.context.new_class = new_class
+
+        parents = [b for b in bases if isinstance(b, cls)]
+        if not parents:
+            return cls.context.new_class
+
+        cls._set_up_config_getter()
+        cls._check_states_enum()
+        cls._check_if_states_are_strings()
+        cls._set_up_translator()
+        cls._calculate_state_name()
+        cls._check_state_value()
+        cls._add_standard_attributes()
+        cls._generate_standard_transitions()
+        cls._generate_standard_methods()
+        cls._generate_named_checkers()
+        cls._generate_named_transitions()
+        cls._add_new_methods()
+        cls._set_complete_option()
+        cls._complete_meta_for_new_class()
+
+        new_class = cls.context.new_class
+        del cls.context
+        return new_class
+
+    @classmethod
+    def _set_up_context(cls):
+        """Create context to keep all needed variables in."""
+        cls.context = AttributeDict()
+        cls.context.new_meta = {}
+        cls.context.new_transitions = {}
+        cls.context.new_methods = {}
+
+    @classmethod
+    def _check_states_enum(cls):
+        """Check if states enum exists and is proper one."""
+        states_enum_name = cls.context.get_config("states_enum_name")
+        try:
+            cls.context["states_enum"] = getattr(cls.context.new_class, states_enum_name)
+        except AttributeError as err:
+            raise ValueError("No states enum given!") from err
+
+        proper = True
+        try:
+            if not issubclass(cls.context.states_enum, Enum):
+                proper = False
+        except TypeError:
+            proper = False
+
+        if not proper:
+            raise ValueError("Please provide enum instance to define available states.")
+
+    @classmethod
+    def _check_if_states_are_strings(cls):
+        """Check if all states are strings."""
+        for item in list(cls.context.states_enum):
```

Comment:
unnecessary conversion to list (?)

Reply:
Obsolete: Vendored entire package using `vendoring`.